prune/prune.go

```diff
@@ -16,184 +16,209 @@ package prune
-		}
-		if err != nil {
-			return err
-		}
+// StrategyFunc takes a list of resources and returns the subset to prune.
+type StrategyFunc func(ctx context.Context, objs []client.Object) ([]client.Object, error)
 
-		err = config.removeResources(ctx, resourcesToRemove)
-		if err != nil {
-			return err
-		}
-	}
+// IsPrunableFunc is a function that checks the data of an object to see whether or not it is safe to prune it.
+// It should return `nil` if it is safe to prune, `Unprunable` if it is unsafe, or another error.
+// It should safely assert the object is the expected type, otherwise it might panic.
+type IsPrunableFunc func(obj client.Object) error
 
-	log.V(1).Info("Prune completed")
+// PrunerOption configures the pruner.
+type PrunerOption func(p *Pruner)
 
-	return nil
+// WithRegistry can be used to set the Registry field when configuring a Pruner
+func WithRegistry(registry Registry) PrunerOption {
+	return func(p *Pruner) {
+		p.Registry = registry
+	}
 }
 
-// containsString checks if a string is present in a slice
-func containsString(s []string, str string) bool {
-	for _, v := range s {
-		if v == str {
-			return true
-		}
+// WithStrategy can be used to set the Strategy field when configuring a Pruner
+func WithStrategy(strategy StrategyFunc) PrunerOption {
+	return func(p *Pruner) {
+		p.Strategy = strategy
 	}
+}
 
-	return false
+// WithNamespace can be used to set the Namespace field when configuring a Pruner
+func WithNamespace(namespace string) PrunerOption {
+	return func(p *Pruner) {
+		p.Namespace = namespace
+	}
 }
 
-// containsName checks if a string is present in a ResourceInfo slice
-func containsName(s []ResourceInfo, str string) bool {
-	for _, v := range s {
-		if v.Name == str {
-			return true
-		}
+// WithLogger can be used to set the Logger field when configuring a Pruner
+func WithLogger(logger logr.Logger) PrunerOption {
+	return func(p *Pruner) {
+		p.Logger = logger
 	}
+}
 
-	return false
+// WithLabels can be used to set the Labels field when configuring a Pruner
+func WithLabels(labels map[string]string) PrunerOption {
+	return func(p *Pruner) {
+		p.Labels = labels
+	}
 }
-func (config Config) validate() (err error) {
 
-	if config.CustomStrategy == nil && config.Strategy.Mode == CustomStrategy {
-		return fmt.Errorf("custom strategies require a strategy function to be specified")
+// WithGVK can be used to set the GVK field when configuring a Pruner
+func WithGVK(gvk schema.GroupVersionKind) PrunerOption {
+	return func(p *Pruner) {
+		p.GVK = gvk
 	}
+}
 
-	if len(config.Namespaces) == 0 {
-		return fmt.Errorf("namespaces are required")
+// NewPruner returns a pruner that uses the given strategy to prune objects.
+func NewPruner(prunerClient client.Client, opts ...PrunerOption) Pruner {
+	pruner := Pruner{
+		Registry: defaultRegistry,
+		Client:   prunerClient,
+		Logger:   Logger(context.Background(), Pruner{}),
+		GVK:      corev1.SchemeGroupVersion.WithKind("Pod"),
```

Comment:
I am a bit uneasy with the defaulting. If GVK, namespaces and labels are not set by the caller, who is relying on the default values, all the Pods of all the namespaces are candidates for deletion. DefaultPodIsPrunable will thankfully limit that to Pods, which have succeed.

Reply:
Thanks for bringing this up, @ryantking and I spoke offline a little bit about this yesterday and decided that it would be best to require the GVK be passed in as a parameter to the `NewPruner` function. I am currently working on making this change now.

Reply:
done in https://github.com/operator-framework/operator-lib/pull/105/commits/67688741005b777c097f0c3952c3d6c010d15db9